google/cloud/aiplatform/models.py

```diff
@@ -215,10 +244,548 @@ def upload(
+        """
+        if endpoint_obj is None:
+            display_name = self._gca_resource.display_name + "_endpoint"
+            endpoint_obj = Endpoint.create(display_name=display_name)
+
+            # TODO(b/171631203) queue deploy instead of block
+            @retry.Retry(deadline=timeout)
+            def endpoint_exist():
+                if endpoint_obj._gca_resource is None:
+                    raise Exception("Endpoint not yet created.")
+
+            endpoint_exist()
+
+        endpoint_obj.deploy(
+            self,
+            deployed_model_display_name=deployed_model_display_name,
+            traffic_percentage=traffic_percentage,
+            machine_type=machine_type,
+            min_replica_count=min_replica_count,
+            max_replica_count=max_replica_count,
+            metadata=metadata,
+        )
+
+        return endpoint_obj
+
+
+class Endpoint(base.AiPlatformResourceNoun):
+
+    client_class = EndpointServiceClient
+    _is_client_prediction_client = False
+
+    def __init__(
+        self,
+        endpoint_name: Optional[str] = None,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ):
+        """Retrieves an endpoint resource.
+
+        Args:
+            endpoint_name (str):
+                Required. A fully-qualified endpoint resource name or endpoint ID.
+                Example: "projects/123/locations/us-central1/endpoints/456" or
+                "456" when project and location are initialized or passed.
+            project (str):
+                Optional. Project to retrieve endpoint from. If not set, project
+                set in aiplatform.init will be used.
+            location (str):
+                Optional. Location to retrieve endpoint from. If not set, location
+                set in aiplatform.init will be used.
+            credentials (auth_credentials.Credentials):
+                Optional. Custom credentials to use to upload this model. Overrides
+                credentials set in aiplatform.init.
+        """
+
+        super().__init__(project=project, location=location, credentials=credentials)
+        if endpoint_name:
+            self._gca_resource = _get_endpoint(endpoint_name)
+
+    def _get_endpoint(self, endpoint_name: str) -> gca_endpoint.Endpoint:
+        """Gets the endpoint from AI Platform.
+
+        Args:
+            endpoint_name (str):
+                Required. The name of the endpoint to retrieve.
+        Returns:
+            endpoint (gca_endpoint.Endpoint):
+                Managed endpoint resource.
+        """
+
+        # Fully qualified endpoint name, i.e. "projects/.../locations/.../endpoints/12345"
+        valid_name = utils.extract_fields_from_resource_name(
+            resource_name=endpoint_name, resource_noun="endpoints"
+        )
+
+        # Partial endpoint name (i.e. "12345") with known project and location
+        if (
```

Comment:
if this logic is the same as the other resource nouns, refactor it.

Reply:
i'll look to add this to utils unless there's objection